### PR TITLE
PERF: uses bincount instead of hash table in categorical value counts

### LINKED ISSUE #10804: PERF: categorical value_counts can be much faster
The internal impl of ``Categorical.value_counts`` should just do this. I think it is factorizing multiple times when it is not necessary.

```
In [32]: np.random.seed(1234)

In [33]: n = 500000

In [34]: u = int(0.1*n)

In [35]: arr = [ "s%04d" % i for i in np.random.randint(0,u,size=n) ]

In [36]: c = pd.Series(arr).astype('category')                              

In [37]: result1 = Series(np.arange(len(c.cat.categories)),c.cat.categories).map(c.cat.codes.value_counts()).order(ascending=False)

In [38]: result2 = c.value_counts()

In [39]: %timeit Series(np.arange(len(c.cat.categories)),c.cat.categories).map(c.cat.codes.value_counts()).order(ascending=False)
100 loops, best of 3: 17.2 ms per loop

In [40]: %timeit c.value_counts()
10 loops, best of 3: 62.3 ms per loop

In [41]: result1.equals(result2)
Out[41]: True
```

### PR DESCRIPTION
**Pull request by behzadnouri**
closes https://github.com/pydata/pandas/issues/10804

``` ipython
In [1]: np.random.seed(2718281)

In [2]: n = 500000

In [3]: u = int(0.1*n)

In [4]: arr = ["s%04d" % i for i in np.random.randint(0, u, size=n)]

In [5]: ts = pd.Series(arr).astype('category')

In [6]: %timeit ts.value_counts()
10 loops, best of 3: 82.7 ms per loop
```

on branch:

``` ipython
In [6]: %timeit ts.value_counts()
10 loops, best of 3: 31.3 ms per loop
```

### CODE REVIEW

**Comment by jreback** [comment]
I think the soln in the issues is faster than this no?

**Comment by behzadnouri** [comment]
I get
```ipython
In [9]: %timeit Series(np.arange(len(ts.cat.categories)),ts.cat.categories).map(ts.cat.codes.value_counts()).order(ascending=False)
10 loops, best of 3: 28.3 ms per loop
```
but this does not check for nulls, and the index is not categorical.

**Comment by behzadnouri** [comment]
with `dropna=False` the branch performs better:

```ipython
In [9]: %timeit ts.value_counts(dropna=True)
10 loops, best of 3: 32.3 ms per loop

In [10]: %timeit ts.value_counts(dropna=False)
10 loops, best of 3: 25.7 ms per loop

In [11]: %timeit Series(np.arange(len(ts.cat.categories)),ts.cat.categories).map(ts.cat.codes.value_counts()).order(ascending=False) 
10 loops, best of 3: 29.5 ms per loop
```

**Comment by jreback** [comment]
wow, this does even better!

```
In [7]: %timeit ts.value_counts(dropna=True)
100 loops, best of 3: 11 ms per loop

In [8]: %timeit ts.value_counts(dropna=False)
100 loops, best of 3: 9.53 ms per loop

In [9]: %timeit Series(np.arange(len(ts.cat.categories)),ts.cat.categories).map(ts.cat.codes.value_counts()).order(ascending=False) 
100 loops, best of 3: 17.4 ms per loop
```

**Comment by jreback** [comment]
ping when green

**Comment by jorisvandenbossche** [comment]
Maybe worth adding a benchmark?

**Comment by behzadnouri** [comment]
I will add benchmark later today

**Comment by behzadnouri** [comment]
added the benchmark, all green.

File: asv_bench/benchmarks/categoricals.py [hunk 1]
```diff
@@ -8,4 +8,18 @@ def setup(self):
...
+
+    def setup(self):
+        n = 500000
+        np.random.seed(2718281)
+        arr = ['s%04d' % i for i in np.random.randint(0, n // 10, size=n)]
+        self.ts = Series(arr).astype('category')
+
+    def time_value_counts(self):
+        self.ts.value_counts(dropna=True)
+        self.ts.value_counts(dropna=False)
```
Comment (jreback) [inline asv_bench/benchmarks/categoricals.py]: These should have only 1 action per timing function (so make 2 functions)

Comment (behzadnouri) [inline asv_bench/benchmarks/categoricals.py]: why should it be only 1 action?

Comment (jorisvandenbossche) [inline asv_bench/benchmarks/categoricals.py]: You get a timing per function. So if you want to track performance of both with dropna True and False, it has to be in two functions.

Comment (behzadnouri) [inline asv_bench/benchmarks/categoricals.py]: added separate calls

**Comment by jreback** [comment]
thank you sir!

**Pull request merged**